File: examples/rco_walk.ts
// One worked territory on RCO-A2A: resolve, verify, and walk the record chain.
// Walk: resolve_compliance(gtin:00990832300082, IT-ECO-10060) -> verify the ES256 signature against the dpuone
// keyring -> get_record on the record and down its `supersedes` chain -> the same chain walk on the rail's one
// supersession pair (for the shape of a chain with depth) -> print -> stop. Node 24, no dependencies.
// Usage: node rco_walk.ts [OBJECT_ID] [JURISDICTION]
import { createPublicKey, verify as cryptoVerify } from "node:crypto";

const MCP = "https://mcp.rco-a2a.ai/mcp";
const KEYRING = "https://dpuone.ai/.well-known/jwks.json";
const UA = "rco-a2a-kit/1.0 (+https://github.com/greencore-solutions/rco-a2a)";
const OBJECT_ID = process.argv[2] ?? "gtin:00990832300082";
const JURIS = process.argv[3] ?? "IT-ECO-10060";
const CHAIN_DEMO = "rco:gsc:gtin:00990832300105:FR-ECO-10060:2"; // the rail's one supersession, as of 2026-09-09

async function call(name: string, args: Record<string, unknown>, id: number): Promise<any> {
  const r = await fetch(MCP, { method: "POST", headers: { "User-Agent": UA, "Content-Type": "application/json", Accept: "application/json, text/event-stream" },
    body: JSON.stringify({ jsonrpc: "2.0", id, method: "tools/call", params: { name, arguments: args } }) });
  const raw = await r.text();
  const msg = raw.trimStart().startsWith("{") ? JSON.parse(raw) : JSON.parse(raw.split("\n").find((l) => l.startsWith("data:"))!.slice(5));
  if (msg.error) return { error: msg.error };
  if (msg.result.isError) return { error: JSON.parse(msg.result.content[0].text) };
  return msg.result.structuredContent ?? JSON.parse(msg.result.content[0].text);
}

// RFC 8785 canonical JSON for the value space these records use (strings, integers, null, objects, arrays)
function jcs(v: any): string {
  if (v === null || typeof v !== "object") return JSON.stringify(v);
  if (Array.isArray(v)) return "[" + v.map(jcs).join(",") + "]";
  return "{" + Object.keys(v).sort().map((k) => JSON.stringify(k) + ":" + jcs(v[k])).join(",") + "}";
}

const KEYS = (await (await fetch(KEYRING, { headers: { "User-Agent": UA } })).json()).keys as any[];

function verify(rec: any): string {
  const [protectedB64, , sigB64] = rec.signature.split(".");
  const hdr = JSON.parse(Buffer.from(protectedB64, "base64url").toString("utf8"));
  const jwk = KEYS.find((k) => k.kid === hdr.kid);
  if (!jwk) return `FAILED - kid ${hdr.kid} not on the keyring`;
  const { signature: _s, ...body } = rec;
  const input = Buffer.concat([Buffer.from(protectedB64 + ".", "ascii"), Buffer.from(jcs(body), "utf8")]);
  const ok = cryptoVerify("sha256", input, { key: createPublicKey({ key: jwk, format: "jwk" }), dsaEncoding: "ieee-p1363" }, Buffer.from(sigB64, "base64url"));
  return ok ? `verified - ES256, kid ${jwk.kid}` : "FAILED - does not verify";
}

const brief = (rec: any) => `${rec.record_id} | ${rec.signal} ${rec.state} | resolved ${rec.resolved_at} | valid_until ${rec.valid_until} | supersedes ${rec.supersedes}`;

async function walkChain(recordId: string | null, id: number): Promise<number> {
  let depth = 0;
  while (recordId) {
    const rec = await call("get_record", { record_id: recordId }, id++);
    if (rec.error) { console.log(`   get_record(${recordId}) ->`, JSON.stringify(rec.error)); break; }
    depth++;
    console.log(`   get_record -> ${brief(rec)}  [${verify(rec)}]`);
    recordId = rec.supersedes ?? null;
  }
  console.log("   chain depth:", depth);
  return id;
}

console.log(`1 resolve_compliance(${OBJECT_ID}, ${JURIS})`);
const rec = await call("resolve_compliance", { object_id: OBJECT_ID, jurisdiction: JURIS }, 1);
if (rec.error) { console.log("   ->", JSON.stringify(rec.error)); process.exit(1); }
console.log("   ->", brief(rec));
console.log("   issuer", rec.issuer.name, "| rail", rec.issuer.rail, "| key_id", rec.key_id, "| case_study", rec.case_study);
console.log("2 verify against", KEYRING, "->", verify(rec));
console.log("3 get_record down the supersedes chain from", rec.record_id);
const next = await walkChain(rec.record_id, 2);
console.log("4 the rail's one supersession pair, for the shape:", CHAIN_DEMO);
await walkChain(CHAIN_DEMO, next);

File: examples/rco_walk.py
#!/usr/bin/env python3
"""One worked territory on RCO-A2A: resolve, verify, and walk the record chain.

Walk: resolve_compliance(gtin:00990832300082, IT-ECO-10060) -> verify the ES256 signature against the
dpuone keyring -> get_record on the record and down its `supersedes` chain -> the same chain walk on the
rail's one supersession pair (for the shape of a chain with depth) -> print -> stop.
Standard library only; the signature check uses `cryptography` when installed and is reported as skipped otherwise.
Usage: python rco_walk.py [OBJECT_ID] [JURISDICTION]
"""
import base64, json, sys, urllib.request

MCP = "https://mcp.rco-a2a.ai/mcp"
KEYRING = "https://dpuone.ai/.well-known/jwks.json"
UA = "rco-a2a-kit/1.0 (+https://github.com/greencore-solutions/rco-a2a)"
OBJECT_ID = sys.argv[1] if len(sys.argv) > 1 else "gtin:00990832300082"
JURIS = sys.argv[2] if len(sys.argv) > 2 else "IT-ECO-10060"
CHAIN_DEMO = "rco:gsc:gtin:00990832300105:FR-ECO-10060:2"   # the rail's one supersession, as of 2026-09-09


def call(name, args, req_id):
    body = json.dumps({"jsonrpc": "2.0", "id": req_id, "method": "tools/call", "params": {"name": name, "arguments": args}}).encode("utf-8")
    req = urllib.request.Request(MCP, data=body, headers={"User-Agent": UA, "Content-Type": "application/json", "Accept": "application/json, text/event-stream"})
    with urllib.request.urlopen(req, timeout=60) as r:
        raw = r.read().decode("utf-8")
    msg = json.loads(raw) if raw.lstrip().startswith("{") else next(json.loads(l[5:]) for l in raw.splitlines() if l.startswith("data:"))
    if "error" in msg:
        return {"error": msg["error"]}
    res = msg["result"]
    if res.get("isError"):
        return {"error": json.loads(res["content"][0]["text"])}
    return res.get("structuredContent") or json.loads(res["content"][0]["text"])


def b64u(s):
    return base64.urlsafe_b64decode(s + "=" * (-len(s) % 4))


def verify(rec, keys):
    try:
        from cryptography.hazmat.primitives import hashes
        from cryptography.hazmat.primitives.asymmetric import ec
        from cryptography.hazmat.primitives.asymmetric.utils import encode_dss_signature
    except ImportError:
        return "not checked (pip install cryptography)"
    protected_b64, _, sig_b64 = rec["signature"].split(".")
    protected = json.loads(b64u(protected_b64))
    key = next((k for k in keys if k.get("kid") == protected.get("kid")), None)
    if key is None:
        return "FAILED - kid %s not on the keyring" % protected.get("kid")
    pub = ec.EllipticCurvePublicNumbers(int.from_bytes(b64u(key["x"]), "big"), int.from_bytes(b64u(key["y"]), "big"), ec.SECP256R1()).public_key()
    body = {k: v for k, v in rec.items() if k != "signature"}
    signing_input = protected_b64.encode("ascii") + b"." + json.dumps(body, sort_keys=True, separators=(",", ":"), ensure_ascii=False).encode("utf-8")
    raw = b64u(sig_b64)
    try:
        pub.verify(encode_dss_signature(int.from_bytes(raw[:32], "big"), int.from_bytes(raw[32:], "big")), signing_input, ec.ECDSA(hashes.SHA256()))
        return "verified - ES256, kid %s" % key["kid"]
    except Exception:
        return "FAILED - does not verify"


def brief(rec):
    return "%s | %s %s | resolved %s | valid_until %s | supersedes %s" % (
        rec["record_id"], rec["signal"], rec["state"], rec["resolved_at"], rec["valid_until"], rec.get("supersedes"))


def walk_chain(record_id, req_id):
    depth = 0
    while record_id:
        rec = call("get_record", {"record_id": record_id}, req_id)
        req_id += 1
        if "error" in rec:
            print("   get_record(%s) -> %s" % (record_id, rec["error"]))
            break
        depth += 1
        print("   get_record -> %s  [%s]" % (brief(rec), verify(rec, KEYS)))
        record_id = rec.get("supersedes")
    print("   chain depth:", depth)
    return req_id


KEYS = json.load(urllib.request.urlopen(urllib.request.Request(KEYRING, headers={"User-Agent": UA}), timeout=60))["keys"]

print("1 resolve_compliance(%s, %s)" % (OBJECT_ID, JURIS))
rec = call("resolve_compliance", {"object_id": OBJECT_ID, "jurisdiction": JURIS}, 1)
if "error" in rec:
    print("   ->", rec["error"])
    raise SystemExit(1)
print("   ->", brief(rec))
print("   issuer", rec["issuer"]["name"], "| rail", rec["issuer"]["rail"], "| key_id", rec["key_id"], "| case_study", rec.get("case_study"))
print("2 verify against", KEYRING, "->", verify(rec, KEYS))
print("3 get_record down the supersedes chain from", rec["record_id"])
nxt = walk_chain(rec["record_id"], 2)
print("4 the rail's one supersession pair, for the shape:", CHAIN_DEMO)
walk_chain(CHAIN_DEMO, nxt)
